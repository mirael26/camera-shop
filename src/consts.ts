export const AppUrl = {
  Main: '/',
  Catalog: '/catalog',
  Page: '/page_',
  Product: '/product',
  Cart: '/cart',
  NotFound: '/404',
  UnknownError: '/unknown_error',
  ServerUnavailable: '/server_unavailable',
} as const;

export const ApiUrl = {
  Promo: '/promo',
  Products: '/cameras',
  Reviews: '/reviews',
  Similar: '/similar',
  Promocode: '/coupons',
  Order: '/orders',
} as const;

export const Modal = {
  AddToCart: 'addToCartModal',
  DeleteFromCart: 'deleteFromCartModal',
  Review: 'reviewModal',
  OrderSuccess: 'orderSuccessModal',
} as const;

export const Tab = {
  Features: 'features',
  Description: 'description',
} as const;

export const Month = {
  0: 'января',
  1: 'февраля',
  2: 'марта',
  3: 'апреля',
  4: 'мая',
  5: 'июня',
  6: 'июля',
  7: 'августа',
  8: 'сентября',
  9: 'октября',
  10: 'ноября',
  11: 'декабря',
} as const;

export const Param = {
  Sort: 'sort',
  Order: 'order',
  Page: 'page',
  Tab: 'tab',
  PriceMin: 'price_min',
  PriceMax: 'price_max',
  Category: 'category',
  Level: 'level',
  Type: 'type',
} as const;

export const SortType = {
  Price: 'price',
  Rating: 'rating',
} as const;

export const SortOrder = {
  Asc: 'asc',
  Desc: 'desc',
} as const;

export const Category = {
  Camera: 'Фотоаппарат',
  VideoCamera: 'Видеокамера',
} as const;

export const Level = {
  Beginner: 'Нулевой',
  Amateur: 'Любительский',
  Professional: 'Профессиональный',
} as const;

export const CameraType = {
  Digital: 'Цифровая',
  Film: 'Плёночная',
  Instant: 'Моментальная',
  Collectible: 'Коллекционная',
} as const;
