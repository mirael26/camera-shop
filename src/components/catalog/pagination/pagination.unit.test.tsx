import { getByText, screen } from '@testing-library/react';
import { useNavigate, useParams } from 'react-router-dom';
import { AppUrl } from '../../../consts';
import { renderTestApp } from '../../../test/helpers/render-test-app';
import { renderWithReduxAndRouter } from '../../../test/helpers/render-with-redux-and-router';
import Pagination from './pagination';

const noop = () => {};
const spyNavigate = jest.fn();

jest.mock('react-router-dom', () => {
  return {
    ...jest.requireActual('react-router-dom'),
    useParams: jest.fn(),
    useNavigate: jest.fn(),
  }
});

describe('Pagination', () => {
  beforeEach(() => {
    jest.mocked(useParams).mockReturnValue({ page: '1' });
    jest.mocked(useNavigate).mockReturnValue(spyNavigate);
  });

  test('Renders pages-buttons correctly', () => {
    renderWithReduxAndRouter(<Pagination pageCount={3} changeCurrentPage={noop}/>);
    const buttons = screen.getAllByTestId('pagination-item');
    expect(buttons).toHaveLength(3);
  });

  test('Renders prev/next button correctly on first page', () => {
    renderWithReduxAndRouter(<Pagination pageCount={3} changeCurrentPage={noop}/>);
    expect(screen.queryByText(/Назад/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/Далее/i)).toBeInTheDocument();
  });

  test('Renders prev/next button correctly on last page', () => {
    jest.mocked(useParams).mockReturnValue({ page: '3' });

    renderWithReduxAndRouter(<Pagination pageCount={3} changeCurrentPage={noop}/>);

    expect(screen.queryByText(/Назад/i)).toBeInTheDocument();
    expect(screen.queryByText(/Далее/i)).not.toBeInTheDocument();
  });

  test('Renders active link correctly', () => {
    jest.mocked(useParams).mockReturnValue({ page: '3' });

    renderWithReduxAndRouter(<Pagination pageCount={3} changeCurrentPage={noop}/>);
    const pagination = screen.getByTestId('pagination');
    expect(getByText<HTMLLinkElement>(pagination, /2/i)).not.toHaveClass('pagination__link--active');
    expect(getByText<HTMLLinkElement>(pagination, /3/i)).toHaveClass('pagination__link--active');
  });

  test('Pushes page number to url once after mount', () => {
    jest.mocked(useParams).mockReturnValue({});
    renderWithReduxAndRouter(<Pagination pageCount={3} changeCurrentPage={noop}/>, { route: AppUrl.Catalog});

    expect(spyNavigate).toHaveBeenCalledWith(`${AppUrl.Catalog}${AppUrl.Page}1`, {"replace": true});
  });

  test('If page incorrect redirect to 404', () => {
    jest.mocked(useParams).mockReturnValue({ page: '4' });
    renderWithReduxAndRouter(<Pagination pageCount={3} changeCurrentPage={noop}/>, { route: `${AppUrl.Catalog}${AppUrl.Page}4`});

    expect(spyNavigate).toHaveBeenCalledWith(AppUrl.NotFound);
  });
});
