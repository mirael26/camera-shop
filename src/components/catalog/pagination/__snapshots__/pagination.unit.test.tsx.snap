// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="pagination"
        data-testid="pagination"
      >
        <ul
          class="pagination__list"
        >
          <li
            class="pagination__item"
            data-testid="pagination-item"
          >
            <a
              class="pagination__link pagination__link--active"
              href="/catalog/page_1"
            >
              1
            </a>
          </li>
          <li
            class="pagination__item"
            data-testid="pagination-item"
          >
            <a
              class="pagination__link"
              href="/catalog/page_2"
            >
              2
            </a>
          </li>
          <li
            class="pagination__item"
            data-testid="pagination-item"
          >
            <a
              class="pagination__link"
              href="/catalog/page_3"
            >
              3
            </a>
          </li>
          <li
            class="pagination__item"
          >
            <a
              class="pagination__link pagination__link--text"
              href="/catalog/page_2"
            >
              Далее
            </a>
          </li>
        </ul>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="pagination"
      data-testid="pagination"
    >
      <ul
        class="pagination__list"
      >
        <li
          class="pagination__item"
          data-testid="pagination-item"
        >
          <a
            class="pagination__link pagination__link--active"
            href="/catalog/page_1"
          >
            1
          </a>
        </li>
        <li
          class="pagination__item"
          data-testid="pagination-item"
        >
          <a
            class="pagination__link"
            href="/catalog/page_2"
          >
            2
          </a>
        </li>
        <li
          class="pagination__item"
          data-testid="pagination-item"
        >
          <a
            class="pagination__link"
            href="/catalog/page_3"
          >
            3
          </a>
        </li>
        <li
          class="pagination__item"
        >
          <a
            class="pagination__link pagination__link--text"
            href="/catalog/page_2"
          >
            Далее
          </a>
        </li>
      </ul>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
